### DISCUSSION TITLE
Problems with rename refactoring ("F2" in vscode) in the presence of FQN references

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#619] [Q&A] [answered]
I played with a current snapshop of langium (34b37014727694b7f09e969ecc13f74cf735f926), roughly langium 0.4.0.
I started the domain model example (as described in the README.md).

Then I get a selection files, containing:
 * `datatypes.dmodel`
    ```
      datatype String
      datatype Int
      datatype Decimal
      
      package big {             // <--- rename this (2)
          datatype Int          // <--- rename this (1)
          datatype Decimal
      }
      
      package complex {
          datatype Date
      }
    ```
 * `qualified-names.dmodel`
   ```
      //...      
      package foo.bar {
          datatype Complex
      
          entity E2 extends E1 {
              next: E2
              other: baz.E3
              nested: baz.nested.E5
              time: big.Int                  // <-- observe here
          }
      }
      //...
   ```

When renaming `big.Int` (put cursor on `Int` in package `big` and press F2) to `big.Int2` I get a problem in `qualified-names.dmodel`: the reference changes to `time: Int2` instead of `big.Int2`):
```
package foo.bar {
    datatype Complex

    entity E2 extends E1 {
        next: E2
        other: baz.E3
        nested: baz.nested.E5
        time: Int2
    }
}
``` 

I also observed when renaming the package `big` into `big2` (in `datatypes.dmodel`) nothing changes in the file `qualified-names.dmodel`...

I wonder if we could fix/enhance the (custom) references in order to support such refactoring operations?
(I was motivated by the discussion #615 to try this refactoring)

**Comment by msujew** [comment, marked as answer]
Hey @goto40, we already have an issue on this, see https://github.com/langium/langium/issues/236. Someone from our side is taking a look at it in their spare time (see [this branch](https://github.com/langium/langium/tree/if/277-experimental)), but it's work in progress.

In general, FQN references pose a general issue to the LSP protocol, since it's main user (vscode) expects every word to be a single reference/value. This isn't the case with FQN, so some special case has to be taken on the language server side in order for these things to work correctly.
